crates/turborepo-cache/src/cache_archive/restore.rs

```diff
@@ -0,0 +1,1088 @@
+                .link_name()?
+                .expect("symlink without linkname");
+
+            let processed_linkname = canonicalize_linkname(anchor, &processed_name, &linkname)?;
+
+            let source_node = *nodes
+                .entry(processed_sourcename.clone())
+                .or_insert_with(|| graph.add_node(processed_sourcename.clone()));
+            let link_node = *nodes
+                .entry(processed_linkname.clone())
+                .or_insert_with(|| graph.add_node(processed_linkname.clone()));
+
+            graph.add_edge(source_node, link_node, ());
+
+            header_lookup.insert(processed_sourcename, entry.header().clone());
+        }
+
+        let nodes = petgraph::algo::toposort(&graph, None)
+            .map_err(|_| CacheError::CycleDetected(Backtrace::capture()))?;
+
+        for node in nodes {
+            let key = &graph[node];
+
+            let Some(header) = header_lookup.get(key) else {
+                continue
+            };
+            let file = restore_symlink_allow_missing_target(dir_cache, anchor, header)?;
+            restored.push(file);
+        }
+
+        Ok(restored)
+    }
+}
+
+fn restore_entry<T: Read>(
+    dir_cache: &mut CachedDirTree,
+    anchor: &AbsoluteSystemPath,
+    entry: &mut Entry<T>,
+) -> Result<AnchoredSystemPathBuf, CacheError> {
+    let header = entry.header();
+
+    match header.entry_type() {
+        tar::EntryType::Directory => restore_directory(dir_cache, anchor, entry.header()),
+        tar::EntryType::Regular => restore_regular(dir_cache, anchor, entry),
+        tar::EntryType::Symlink => restore_symlink(dir_cache, anchor, entry.header()),
+        ty => Err(CacheError::UnsupportedFileType(ty, Backtrace::capture())),
+    }
+}
+
+pub fn canonicalize_name(name: &Path) -> Result<AnchoredSystemPathBuf, CacheError> {
+    #[allow(unused_variables)]
+    let PathValidation {
+        well_formed,
+        windows_safe,
+    } = check_name(name);
+
+    if !well_formed {
+        return Err(CacheError::MalformedName(
+            name.to_string_lossy().to_string(),
+            Backtrace::capture(),
+        ));
+    }
+
+    #[cfg(windows)]
+    {
+        if !windows_safe {
+            return Err(CacheError::WindowsUnsafeName(
+                name.to_string(),
+                Backtrace::capture(),
+            ));
+        }
+    }
+
+    // There's no easier way to remove trailing slashes in Rust
+    // because `OsString`s are really just `Vec<u8>`s.
+    let no_trailing_slash: PathBuf = name.components().collect();
+
+    // We know this is indeed anchored because of `check_name`,
+    // and it is indeed system because we just split and combined with the
+    // system path separator above
+    Ok(unsafe { AnchoredSystemPathBuf::new_unchecked(no_trailing_slash) })
+}
+
+#[derive(Debug, PartialEq)]
+struct PathValidation {
+    well_formed: bool,
+    windows_safe: bool,
+}
+
+fn check_name(name: &Path) -> PathValidation {
```

Comment:
Yeah I can do that, although if/when we do the UTF-8 conversion, we can just use normal string methods

Reply:
Actually...hmm I'm not sure it's a good idea to wrap this in a `RelativeUnixPath`, because the whole point of this function is to detect whether it's actually a `RelativeUnixPath` or something else.

Reply:
Yeah this might be the simplest way to do this until we move to UTF-8, because `canonicalize_name` needs to return an `AnchoredSystemPathBuf`, which is really a `PathBuf` underneath the hood, and the only way to get a `PathBuf` from `&[u8]` is via a `PathBuf`.